vcd/resource_vcd_snat.go

```diff
@@ -119,23 +124,44 @@ func resourceVcdSNATCreate(d *schema.ResourceData, meta interface{}) error {
 func resourceVcdSNATRead(d *schema.ResourceData, meta interface{}) error {
 	vcdClient := meta.(*VCDClient)
 
-	e, err := vcdClient.GetEdgeGatewayFromResource(d, "edge_gateway")
+	edgeGateway, err := vcdClient.GetEdgeGatewayFromResource(d, "edge_gateway")
 	if err != nil {
 		return fmt.Errorf(errorUnableToFindEdgeGateway, err)
 	}
 
+	// Terraform refresh won't work if Rule was edit in UI. vCD API uses tag elements to map edge gtw Id's
+	// and UI will reset the tag element on update.
+
 	var found bool
 
-	providedNetworkName := d.Get("network_name")
-	if nil != providedNetworkName && providedNetworkName.(string) != "" {
-		for _, r := range e.EdgeGateway.Configuration.EdgeGatewayServiceConfiguration.NatService.NatRule {
-			if r.RuleType == "SNAT" && r.GatewayNatRule.TranslatedIP == d.Get("internal_ip").(string) && r.GatewayNatRule.Interface.Name == d.Get("network_name").(string) {
-				found = true
-				d.Set("external_ip", r.GatewayNatRule.OriginalIP)
+	networkName := d.Get("network_name")
+	if nil != networkName && networkName.(string) != "" {
+		natRule, err := edgeGateway.FetchNatRule(d.Id())
+		if err != nil {
+			d.SetId("")
+			return err
```

Comment:
A bit more context would help the error as this is the one going for user.

Reply:
fixed